src/common/hist_util.h

```diff
@@ -200,19 +295,22 @@ struct GHistIndexMatrix {
   /*! \brief row pointer to rows by element position */
   std::vector<size_t> row_ptr;
   /*! \brief The index data */
-  std::vector<uint32_t> index;
+  Index index;
   /*! \brief hit count of each index */
   std::vector<size_t> hit_count;
   /*! \brief The corresponding cuts */
   HistogramCuts cut;
+  DMatrix* p_fmat_;
+  size_t max_num_bins_;
   // Create a global histogram matrix, given cut
   void Init(DMatrix* p_fmat, int max_num_bins);
-  // get i-th row
-  inline GHistIndexRow operator[](size_t i) const {
-    return {&index[0] + row_ptr[i],
-            static_cast<GHistIndexRow::index_type>(
-                row_ptr[i + 1] - row_ptr[i])};
-  }
+
+  template<typename T>
+  void SetIndexData(T* const index_data, size_t batch_threads, const SparsePage& batch,
```

Comment:
So is `T*` const?

Reply:
yes pointer is constant (and immutable) but the pointed data is not
`T` was replaced by `BinIdxType`, thanks